sqlglot/dialects/athena.py

```diff
@@ -1,31 +1,131 @@
+    if len(raw_tokens) > 0:
+        first_token = raw_tokens[0]
+        if first_token.token_type == TokenType.CREATE:
+            # CREATE is Hive (except for CREATE VIEW and CREATE TABLE... AS SELECT)
+            return not any(t.token_type in (TokenType.VIEW, TokenType.SELECT) for t in raw_tokens)
+
+        # ALTER and DROP are Hive
+        return first_token.token_type in (TokenType.ALTER, TokenType.DROP)
+    return False
+
+
+def _generate_as_hive(expression: exp.Expression) -> bool:
+    if isinstance(expression, exp.Create):
+        if expression.kind == "TABLE":
+            properties: t.Optional[exp.Properties] = expression.args.get("properties")
+            if properties and properties.find(exp.ExternalProperty):
+                return True  # CREATE EXTERNAL TABLE is Hive
+
+            if not isinstance(expression.expression, exp.Select):
+                return True  # any CREATE TABLE other than CREATE TABLE AS SELECT is Hive
+        else:
+            return expression.kind != "VIEW"  # CREATE VIEW is never Hive but CREATE SCHEMA etc is
+
+    elif isinstance(expression, exp.Alter) or isinstance(expression, exp.Drop):
+        return True  # all ALTER and DROP statements are Hive
+
+    return False
 
 
 class Athena(Trino):
+    """
+    Over the years, it looks like AWS has taken various execution engines, bolted on AWS-specific modifications and then
+    built the Athena service around them.
+
+    Thus, Athena is not simply hosted Trino, it's more like a router that routes SQL queries to an execution engine depending
+    on the query type.
+
+    As at 2024-09-10, assuming your Athena workgroup is configured to use "Athena engine version 3", the following engines exist:
+
+    Hive:
+     - Accepts mostly the same syntax as Hadoop / Hive
+     - Uses backticks to quote identifiers
+     - Has a distinctive DDL syntax (around things like setting table properties, storage locations etc) that is different from Trino
+     - Used for *most* DDL, with some exceptions that get routed to the Trino engine instead:
+        - CREATE [EXTERNAL] TABLE (without AS SELECT)
+        - ALTER
+        - DROP
+
+    Trino:
+      - Uses double quotes to quote identifiers
+      - Used for DDL operations that involve SELECT queries, eg:
+        - CREATE VIEW
+        - CREATE TABLE... AS SELECT
+      - Used for DML operations
+        - SELECT, INSERT, UPDATE, DELETE, MERGE
+
+    The SQLGlot Athena dialect tries to identify which engine a query would be routed to and then uses the parser / generator for that engine
+    rather than trying to create a universal syntax that can handle both types.
+    """
+
     class Tokenizer(Trino.Tokenizer):
+        """
+        The Tokenizer is flexible enough to tokenize queries across both the Hive and Trino engines
+        """
+
         IDENTIFIERS = ['"', "`"]
         KEYWORDS = {
+            **Hive.Tokenizer.KEYWORDS,
             **Trino.Tokenizer.KEYWORDS,
```

Comment:
FYI there's one keyword that maps to a different token between Hive, Trino - `"ROW"`:

```python
>>> from sqlglot.dialects import Hive, Trino
>>> {k: (v, Trino.Tokenizer.KEYWORDS[k]) for k, v in Hive.Tokenizer.KEYWORDS.items() if k in Trino.Tokenizer.KEYWORDS and Trino.Tokenizer.KEYWORDS[k] != Hive.Tokenizer.KEYWORDS[k]}
{'ROW': (<TokenType.ROW: 'ROW'>, <TokenType.STRUCT: 'STRUCT'>)}
>>> {k: (v, Hive.Tokenizer.KEYWORDS[k]) for k, v in Trino.Tokenizer.KEYWORDS.items() if k in Hive.Tokenizer.KEYWORDS and Hive.Tokenizer.KEYWORDS[k] != Trino.Tokenizer.KEYWORDS[k]}
{'ROW': (<TokenType.STRUCT: 'STRUCT'>, <TokenType.ROW: 'ROW'>)}
```

This combination of dicts means we'll always use Trino's override. Not sure if this is fine, just pointing it out to double check for possible unintended side-effects.

Reply:
I think this particular instance is ok because unless i'm mistaken, `ROW` only appears in SELECT queries and Athena uses the Trino syntax for SELECT queries. It's why the Hive tokenizer keywords are listed first and then overridden by the Trino ones.

The higher problem is that I couldn't see an obvious way to determine which tokenizer to use pre-tokenization, however I also couldn't think of a query to add to the tests that exposed this as a problem

Reply:
Ok sounds good, can always revisit later if needed.